enigmahandlers/fieldcache.go

```diff
@@ -0,0 +1,45 @@
+package enigmahandlers
+
+import (
+	"sync"
+
+	"github.com/qlik-oss/enigma-go"
+)
+
+type (
+	// fieldCache
+	fieldCache struct {
+		fieldMap map[string]*enigma.Field
+		mutex    sync.RWMutex
+	}
+)
+
+func NewFieldCache() fieldCache {
+	return fieldCache{fieldMap: make(map[string]*enigma.Field)}
+
+}
+
+func (fc *fieldCache) Lookup(name string) (field *enigma.Field, hit bool) {
+	fc.mutex.RLock()
+	defer fc.mutex.RUnlock()
+	field, hit = fc.fieldMap[name]
+	return
+}
+
+func (fc *fieldCache) LookupWithFallback(fieldName string, fallback func(fieldName string) (*enigma.Field, error)) (*enigma.Field, error) {
```

Comment:
Its meant as some kind of caching function decorator.

It does not always have to fetch, if I understand your question.

The behaviour I am after is: take from cache, or else, fetch and store.

Reply:
Exactly. So the fetch logic could just be in the function rather than sending it in each time. Because right now, the following part:

`func(fieldName string) (*enigma.Field, error) {
					return doc.GetField(ctx, fieldName)
				},`

is duplicated for each function invocation. So instead of a "LookupWithFallback" function more like a "LookupOrFetch"

Reply:
Cool! I see what you mean. I will write such functions to reduce duplication. They will end up outside of the caching classes though.